File: runtime/src/transaction/rwset.rs
//! Read/write set.

/// A coarsened key prefix that represents any key that starts with
/// this prefix.
#[derive(Clone, Debug, Default, PartialEq, Eq, cbor::Encode, cbor::Decode)]
#[cbor(transparent)]
pub struct CoarsenedKey(pub Vec<u8>);

impl AsRef<[u8]> for CoarsenedKey {
    fn as_ref(&self) -> &[u8] {
        &self.0
    }
}

impl From<CoarsenedKey> for Vec<u8> {
    fn from(val: CoarsenedKey) -> Self {
        val.0
    }
}

impl From<Vec<u8>> for CoarsenedKey {
    fn from(v: Vec<u8>) -> CoarsenedKey {
        CoarsenedKey(v)
    }
}

/// A set of coarsened keys.
pub type CoarsenedSet = Vec<CoarsenedKey>;

/// A read/write set.
#[derive(Clone, Debug, Default, PartialEq, Eq, cbor::Encode, cbor::Decode)]
pub struct ReadWriteSet {
    /// Size of the key prefixes (in bytes) used for coarsening the keys.
    pub granularity: u16,
    /// The read set.
    pub read_set: CoarsenedSet,
    /// The write set.
    pub write_set: CoarsenedSet,
}

#[cfg(test)]
mod test {
    use super::*;

    #[test]
    fn test_serialization() {
        let rw_set = ReadWriteSet {
            granularity: 3,
            read_set: vec![b"foo".to_vec().into(), b"bar".to_vec().into()],
            write_set: vec![b"moo".to_vec().into()],
        };

        let enc = cbor::to_vec(rw_set.clone());

        let dec_rw_set: ReadWriteSet = cbor::from_slice(&enc).unwrap();
        assert_eq!(rw_set, dec_rw_set, "serialization should round-trip");
    }
}

File: go/runtime/transaction/rwset.go
package transaction

import (
	"bytes"

	"github.com/oasisprotocol/oasis-core/go/common/cbor"
)

// CoarsenedKey is a coarsened key prefix that represents any key that
// starts with this prefix.
type CoarsenedKey []byte

// CoarsenedSet is a set of coarsened keys.
type CoarsenedSet []CoarsenedKey

// Equal checks whether the coarsened set is equal to another.
func (s CoarsenedSet) Equal(other CoarsenedSet) bool {
	if len(s) != len(other) {
		return false
	}
	for idx, value := range s {
		if !bytes.Equal(value, other[idx]) {
			return false
		}
	}
	return true
}

// ReadWriteSet is a read/write set.
type ReadWriteSet struct {
	// Granularity is the size of the key prefixes (in bytes) used for
	// coarsening the keys.
	Granularity uint16 `json:"granularity"`
	// ReadSet is the read set.
	ReadSet CoarsenedSet `json:"read_set"`
	// WriteSet is the write set.
	WriteSet CoarsenedSet `json:"write_set"`
}

// Equal checks whether the read/write set is equal to another.
func (rw ReadWriteSet) Equal(other *ReadWriteSet) bool {
	return rw.Granularity == other.Granularity && rw.ReadSet.Equal(other.ReadSet) && rw.WriteSet.Equal(other.WriteSet)
}

// Size returns the size (in bytes) of the read/write set.
func (rw ReadWriteSet) Size() uint64 {
	return uint64(len(cbor.Marshal(rw)))
}
